invoker/core/src/main/java/com/google/cloud/functions/invoker/TypedFunctionExecutor.java

```diff
@@ -0,0 +1,195 @@
+    try {
+      funcInstance = clazz.getDeclaredConstructor().newInstance();
+    } catch (Exception e) {
+      throw new RuntimeException(
+          "Class "
+              + clazz.getName()
+              + " must declare a valid default constructor to be usable as a strongly typed"
+              + " function. Could not use constructor: "
+              + e.toString());
+    }
+
+    ConfigurationImpl configuration = new ConfigurationImpl();
+    funcInstance.configure(configuration);
+
+    @SuppressWarnings("unchecked")
+    TypedFunctionExecutor executor =
+        new TypedFunctionExecutor(
+            argType.orElseThrow(), (TypedFunction<Object, Object>) funcInstance, configuration);
+    return executor;
+  }
+
+  /**
+   * Returns the {@code T} of a concrete class that implements {@link BackgroundFunction
+   * BackgroundFunction<T>}. Returns an empty {@link Optional} if {@code T} can't be determined.
+   */
+  static Optional<Type> handlerTypeArgument(Class<? extends TypedFunction<?, ?>> functionClass) {
+    return Arrays.stream(functionClass.getMethods())
+        .filter(method -> method.getName().equals(APPLY_METHOD) && method.getParameterCount() == 1)
+        .map(method -> method.getGenericParameterTypes()[0])
+        .findFirst();
+  }
+
+  /** Executes the user's method, can handle all HTTP type methods. */
+  @Override
+  public void service(HttpServletRequest req, HttpServletResponse res) {
+    HttpRequestImpl reqImpl = new HttpRequestImpl(req);
+    HttpResponseImpl respImpl = new HttpResponseImpl(res);
+    ClassLoader oldContextClassLoader = Thread.currentThread().getContextClassLoader();
+
+    try {
+      Thread.currentThread().setContextClassLoader(func.getClass().getClassLoader());
+      handleRequest(reqImpl, respImpl);
+    } finally {
+      Thread.currentThread().setContextClassLoader(oldContextClassLoader);
+      flushResponse(respImpl);
+    }
+  }
+
+  private void handleRequest(HttpRequest req, HttpResponse resp) {
+    WireFormat format = configuration.getFormat();
+
+    Object reqObj = null;
+    try {
+      reqObj = format.deserialize(req, argType);
+    } catch (Throwable t) {
+      logger.log(Level.SEVERE, "Failed to parse request for " + func.getClass().getName(), t);
+      resp.setStatusCode(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+
+    Object respObj = null;
+    try {
+      respObj = this.func.apply(reqObj);
+    } catch (Throwable t) {
+      logger.log(Level.SEVERE, "Failed to execute " + func.getClass().getName(), t);
+      resp.setStatusCode(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+      return;
+    }
+
+    try {
+      format.serialize(respObj, resp);
+    } catch (Throwable t) {
+      logger.log(Level.SEVERE, "Failed to serialize response for " + func.getClass().getName(), t);
+      resp.setStatusCode(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+      return;
+    }
+  }
+
+  private void flushResponse(HttpResponseImpl respImpl) {
```

Comment:
nit - some DRY with https://github.com/GoogleCloudPlatform/functions-framework-java/blob/master/invoker/core/src/main/java/com/google/cloud/functions/invoker/HttpFunctionExecutor.java#L75

I wonder if some of the logic blocks in these classes can be shared, such as some of the "forClass" method.

Reply:
migrated flush implementation into HttpResponeseImpl which helps a bit, in `forClass` a lot of the logic is validation and uses generics -- it feels like it might add cognitive complexity to try to share the implementation there. WDYT?